src/main/java/gregtech/api/util/PositionedWorldEvent.java

```
@@ -0,0 +1,168 @@
+
+    public PositionedWorldEvent() {}
+
+    public PositionedWorldEvent(World world) {
+        this.world = world;
+    }
+
+    public PositionedWorldEvent(Vec3 position) {
+        this.position = position;
+    }
+
+    public PositionedWorldEvent(T thing) {
+        this.thing = thing;
+    }
+
+    public PositionedWorldEvent(World world, T thing) {
+        this(world);
+        this.thing = thing;
+    }
+
+    public PositionedWorldEvent(World world, Vec3 position) {
+        this(world);
+        this.position = position;
+    }
+
+    public PositionedWorldEvent(Vec3 position, T thing) {
+        this(position);
+        this.thing = thing;
+    }
+
+    public PositionedWorldEvent(T thing, Vec3 position) {
+        this(position, thing);
+    }
+
+    public PositionedWorldEvent(World world, Vec3 position, T thing) {
+        this(world, position);
+        this.thing = thing;
+    }
+
+    public PositionedWorldEvent(World world, T thing, Vec3 position) {
+        this(world, position, thing);
+    }
+
+    private World world;
+    private Vec3 position;
+    private T thing;
+
+    public World getWorld() {
+        return world;
+    }
+
+    public PositionedWorldEvent<T> setWorld(World world) {
+        this.world = world;
+        return this;
+    }
+
+    public Vec3 getPosition() {
+        return position;
+    }
+
+    public PositionedWorldEvent<T> setPosition(Vec3 position) {
+        this.position = position;
+        return this;
+    }
+
+    public PositionedWorldEvent<T> setPosition(double x, double y, double z) {
+        this.position = Vec3.createVectorHelper(x, y, z);
+        return this;
+    }
+
+    public T getThing() {
+        return thing;
+    }
+
+    public PositionedWorldEvent<T> setThing(T thing) {
+        this.thing = thing;
+        return this;
+    }
+
+    public void spawnParticle(double motionX, double motionY, double motionZ) {
+        if (position == null || !(thing instanceof String) || world == null)
```

I'd like it to throw IllegalArgumentException (or some more appropriate exception class that you can think of) if any of these checks fail. It signifies an error in someone's code. Silently ignoring is just sweeping bugs under the rug. Same for below methods.